authors/apps/articles/views.py

```diff
@@ -84,4 +99,89 @@ def destroy(self, request, pk=None):
         article = get_object_or_404(queryset, pk=pk)
         article.delete()
         return Response({"message": "article deleted successfully"},
-                        status=status.HTTP_200_OK)
+                        status=status.HTTP_204_NO_CONTENT)
+
+
+class RatingAPIView(GenericAPIView):
+    queryset = Rating.objects.all()
+    serializer_class = RatingSerializer
+    permission_class = (IsAuthenticatedOrReadOnly,)
+
+    def get_article(self, slug):
+        """Returns an article based on slug."""
+        article = Article.objects.all().filter(slug=slug).first()
+        return article
+
+    def post(self, request, slug):
+        """POST request to rate an article."""
+        rating = request.data
+        article = self.get_article(slug)
+
+        if not article:
+            raise ValidationError(
+                detail={'message': 'Article not found'}
+            )
+        if request.user.is_authenticated is False:
+            return Response({
+                "message": "Please login to rate an article"
+            }, status=status.HTTP_403_FORBIDDEN)
+
+        if request.user.id == article.author.id:
+            return Response({
+                "message": "You cannot rate your own article"
+            }, status=status.HTTP_403_FORBIDDEN)
+
+        try:
+            current_rating = Rating.objects.get(
+                user=request.user.id,
+                article=article
+            )
+            serializer = self.serializer_class(
+                current_rating, data=rating)
+        except Rating.DoesNotExist:
+            serializer = self.serializer_class(data=rating)
+
+        serializer.is_valid(raise_exception=True)
+        serializer.save(user=request.user, article=article)
+
+        return Response({
+            'message': 'Rating submitted sucessfully',
+            'data': serializer.data
+        }, status=status.HTTP_201_CREATED)
+
+    def get(self, request, slug):
+        """Get request for an article ratings."""
+        article = self.get_article(slug)
+        rating = None
+
+        if not article:
+            raise ValidationError(
+                detail={'message': 'Article not found'}
+            )
+
+        if request.user.is_authenticated:
+            try:
+                rating = Rating.objects.get(user=request.user, article=article)
+            except Rating.DoesNotExist:
+                rating = None
+
+        if rating is None:
+            avg = Rating.objects.filter(
+                article=article).aggregate(Avg('user_rating')
+                                           )
+            average_rating = avg['user_rating__avg']
+            if avg["user_rating__avg"] is None:
+                average_rating = 0
+
+            if request.user.is_authenticated is False:
```

Comment:
Should be handled by the `permissions_class`

Reply:
this route allows both authenticated and unauthenticated users to access it but returns different results for each scenario

Reply:
Unauthenticated users can read an article and also see an articles ratings